[Bug]: version 3.7.0 is has a pixelation bug

### Checklist

- [X] I have read the appropriate section in the [contributing guidelines](https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md)
- [X] I believe this issue is a problem with polybar itself and not a misconfiguration on my part
- [X] I have searched for other open and closed [issues](https://github.com/polybar/polybar/issues?q=is%3Aissue) that may have already reported this problem
- [X] I have checked the [known issues](https://github.com/polybar/polybar/wiki/Known-Issues) page for this problem.
- [X] I have followed the [debugging guide](https://github.com/polybar/polybar/wiki/Debugging-your-Config) to narrow down the problem to a minimal config.

### Steps to reproduce

1. I simply launched my standard `polybar` after upgrading from 3.6.3 -> 3.7.0 and it became surrounded with colourful pixels like this:
![image](https://github.com/polybar/polybar/assets/24229496/f7bf2adf-609a-43ba-bb50-b4573a7fbf63)

Note, that I'm running Arch Linux with i3 on an asus laptop with an intel video card through a Dell WD19TB dock connected to 2 external monitors.
The issue wasn't there when I disconnected the dock and only used my laptop screen.

I tried downgrading polybar back to 3.6.3 and the issue was immediately fixed.

For anyone on arch, you can downgrade to your cached version by running something like:
`sudo pacman -U file:///var/cache/pacman/pkg/polybar-3.6.3-3-x86_64.pkg.tar.zst`


### Minimal config

```dosini
I tested without any custom configuration and the issue was there.
```


### Polybar log

_No response_

### Expected behavior

I wasn't expecting to see any colorful pixels.

### Actual behavior

My polybars, on all 3 screens, were surrounded by colorful pixels.

### Window Manager and Version

i3 version 4.23

### Linux Distribution

Arch Linux

### Polybar version

```text
polybar 3.7.0

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) +composite +xkb +xrm +xcursor

Build type: Release
Compiler: /usr/bin/c++
Compiler flags: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -g -ffile-prefix-map=/build/polybar/src=/usr/src/debug/polybar -flto=auto -O3 -DNDEBUG -Wall -Wextra -Wpedantic -Wdeprecated-copy-dtor -Wsuggest-override
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -flto=auto -Wall -Wextra -Wpedantic -Wdeprecated-copy-dtor -Wsuggest-override  -Wall -Wextra -Wpedantic -Wdeprecated-copy-dtor -Wsuggest-override
```


### Additional Context / Screenshots

I should also note that I repeated this test on 2 different laptops and 2 different Thunderbolt docking stations, and the bug persisted in all cases.

I'm able to build 3.7.0 successfully, but not 3.6.3.
These are the logs I get when I run `./build.sh -f` on `a962a265 (HEAD, tag: 3.6.3) Version 3.6.3`

```
[ 77%] Building CXX object bin/CMakeFiles/poly.dir/x11/window.cpp.o
[ 78%] Building CXX object bin/CMakeFiles/poly.dir/x11/winspec.cpp.o
/home/mike/temp/polybar/src/x11/tray_manager.cpp: In member function ‘void polybar::tray_manager::setup(const polybar::bar_settings&)’:
/home/mike/temp/polybar/src/x11/tray_manager.cpp:70:17: warning: possibly dangling reference to a temporary [-Wdangling-reference]
   70 |   const config& conf = config::make();
      |                 ^~~~
/home/mike/temp/polybar/src/x11/tray_manager.cpp:70:36: note: the temporary was destroyed at the end of the full expression ‘polybar::config::make(std::__cxx11::basic_string<char>(((const char*)""), std::allocator<char>()), std::__cxx11::basic_string<char>(((const char*)""), std::allocator<char>()))’
   70 |   const config& conf = config::make();
      |                        ~~~~~~~~~~~~^~
[ 79%] Building CXX object bin/CMakeFiles/poly.dir/x11/xembed.cpp.o
[ 80%] Building CXX object bin/CMakeFiles/poly.dir/adapters/alsa/control.cpp.o
[ 80%] Building CXX object bin/CMakeFiles/poly.dir/adapters/alsa/mixer.cpp.o
[ 81%] Building CXX object bin/CMakeFiles/poly.dir/modules/alsa.cpp.o
In file included from /home/mike/temp/polybar/include/adapters/alsa/generic.hpp:37,
                 from /home/mike/temp/polybar/src/adapters/alsa/mixer.cpp:3:
/usr/include/alsa/control.h:417:47: error: ‘snd_ump_endpoint_info_t’ has not been declared
  417 | int snd_ctl_ump_endpoint_info(snd_ctl_t *ctl, snd_ump_endpoint_info_t *info);
      |                                               ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/alsa/control.h:418:44: error: ‘snd_ump_block_info_t’ has not been declared
  418 | int snd_ctl_ump_block_info(snd_ctl_t *ctl, snd_ump_block_info_t *info);
      |                                            ^~~~~~~~~~~~~~~~~~~~
In file included from /home/mike/temp/polybar/include/adapters/alsa/generic.hpp:37,
                 from /home/mike/temp/polybar/src/adapters/alsa/control.cpp:2:
/usr/include/alsa/control.h:417:47: error: ‘snd_ump_endpoint_info_t’ has not been declared
  417 | int snd_ctl_ump_endpoint_info(snd_ctl_t *ctl, snd_ump_endpoint_info_t *info);
      |                                               ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/alsa/control.h:418:44: error: ‘snd_ump_block_info_t’ has not been declared
  418 | int snd_ctl_ump_block_info(snd_ctl_t *ctl, snd_ump_block_info_t *info);
      |                                            ^~~~~~~~~~~~~~~~~~~~
make[2]: *** [bin/CMakeFiles/poly.dir/build.make:1126: bin/CMakeFiles/poly.dir/adapters/alsa/control.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [bin/CMakeFiles/poly.dir/build.make:1140: bin/CMakeFiles/poly.dir/adapters/alsa/mixer.cpp.o] Error 1
In file included from /home/mike/temp/polybar/include/adapters/alsa/generic.hpp:37,
                 from /home/mike/temp/polybar/src/modules/alsa.cpp:4:
/usr/include/alsa/control.h:417:47: error: ‘snd_ump_endpoint_info_t’ has not been declared
  417 | int snd_ctl_ump_endpoint_info(snd_ctl_t *ctl, snd_ump_endpoint_info_t *info);
      |                                               ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/alsa/control.h:418:44: error: ‘snd_ump_block_info_t’ has not been declared
  418 | int snd_ctl_ump_block_info(snd_ctl_t *ctl, snd_ump_block_info_t *info);
      |                                            ^~~~~~~~~~~~~~~~~~~~
In file included from /home/mike/temp/polybar/include/modules/meta/event_module.hpp:3,
                 from /home/mike/temp/polybar/include/modules/alsa.hpp:3,
                 from /home/mike/temp/polybar/src/modules/alsa.cpp:1:
/home/mike/temp/polybar/include/modules/meta/base.inl: In instantiation of ‘polybar::modules::module<Impl>::module(const polybar::bar_settings&, std::string) [with Impl = polybar::modules::alsa_module; std::string = std::__cxx11::basic_string<char>]’:
/home/mike/temp/polybar/src/modules/alsa.cpp:18:18:   required from here
/home/mike/temp/polybar/include/modules/meta/base.hpp:202:19: warning: possibly dangling reference to a temporary [-Wdangling-reference]
  202 |     const config& m_conf;
      |                   ^~~~~~
In file included from /home/mike/temp/polybar/src/modules/alsa.cpp:9:
/home/mike/temp/polybar/include/modules/meta/base.inl:22:28: note: the temporary was destroyed at the end of the full expression ‘polybar::config::make(std::__cxx11::basic_string<char>(((const char*)""), std::allocator<char>()), std::__cxx11::basic_string<char>(((const char*)""), std::allocator<char>()))’
   22 |       , m_conf(config::make())
      |                ~~~~~~~~~~~~^~
make[2]: *** [bin/CMakeFiles/poly.dir/build.make:1154: bin/CMakeFiles/poly.dir/modules/alsa.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:520: bin/CMakeFiles/poly.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
** Failed to build project
```

--- a/.clang-tidy
+++ b/.clang-tidy
@@ -40,6 +40,7 @@ CheckOptions:
   readability-identifier-naming.MemberCase: lower_case
   readability-identifier-naming.ProtectedMemberPrefix: 'm_'
   readability-identifier-naming.PrivateMemberPrefix: 'm_'
+  readability-simplify-boolean-expr.SimplifyDeMorgan: false
 
 HeaderFilterRegex: ''
 WarningsAsErrors: ''

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `internal/backlight`: Module could display the literal `%percentage%` token if the backlight reports a value of 0 at startup ([`#3081`](https://github.com/polybar/polybar/pull/3081)) by [@unclechu](https://github.com/unclechu)
 - `internal/tray`: Fix crash during restarting, when tray icons were not removed proberly ([`#3111`](https://github.com/polybar/polybar/issues/3111), [`#3112`](https://github.com/polybar/polybar/pull/3112))
 - `custom/ipc`: Module would display the literal `%output%` token before the initial hook finished executing ([`#3131`](https://github.com/polybar/polybar/issues/3131), [`#3140`](https://github.com/polybar/polybar/pull/3140))
+- renderer: Pseudo-transparency rendering artifacts when wallpaper does not fill entire screen ([`#3096`](https://github.com/polybar/polybar/pull/3096), [`#3041`](https://github.com/polybar/polybar/issues/3041))
 
 ## [3.7.1] - 2023-11-27
 ### Build

--- a/src/x11/background_manager.cpp
+++ b/src/x11/background_manager.cpp
@@ -220,13 +220,34 @@ void bg_slice::copy(xcb_pixmap_t root_pixmap, int depth, xcb_rectangle_t geom, x
   ensure_resources(depth, visual);
   assert(m_pixmap);
 
-  // fill the slice
+  auto pixmap_end_x = int16_t(geom.x + geom.width);
+  auto pixmap_end_y = int16_t(geom.y + geom.height);
+
   auto translated = m_connection.translate_coordinates(m_window, m_connection.screen()->root, m_rect.x, m_rect.y);
-  // Coordinates of the slice in the root pixmap
-  auto src_x = math_util::cap(translated->dst_x, geom.x, int16_t(geom.x + geom.width));
-  auto src_y = math_util::cap(translated->dst_y, geom.y, int16_t(geom.y + geom.height));
+
+  /*
+   * If the slice is not fully contained in the root pixmap, we will be missing at least some background pixels. For
+   * those areas, nothing is copied over and a simple black background is shown.
+   * This can happen when connecting new monitors without updating the root pixmap.
+   */
+  if (!(translated->dst_x >= geom.x && translated->dst_x + m_rect.width <= pixmap_end_x &&
+          translated->dst_y >= geom.y && translated->dst_y + m_rect.height <= pixmap_end_y)) {
+    m_log.err(
+        "background_manager: Root pixmap does not fully cover transparent areas. "
+        "Pseudo-transparency may not fully work and instead just show a black background. "
+        "Make sure you have a wallpaper set on all of your screens");
+  }
+
+  /*
+   * Coordinates of the slice in the root pixmap. The rectangle is capped so that it is contained in the root pixmap to
+   * avoid copying areas not covered by the pixmap.
+   */
+  auto src_x = math_util::cap(translated->dst_x, geom.x, pixmap_end_x);
+  auto src_y = math_util::cap(translated->dst_y, geom.y, pixmap_end_x);
   auto w = math_util::cap(m_rect.width, uint16_t(0), uint16_t(geom.width - (src_x - geom.x)));
   auto h = math_util::cap(m_rect.height, uint16_t(0), uint16_t(geom.height - (src_y - geom.y)));
+
+  // fill the slice
   m_log.trace(
       "background_manager: Copying from root pixmap (0x%x:%d) %dx%d+%d+%d", root_pixmap, depth, w, h, src_x, src_y);
   m_connection.copy_area_checked(root_pixmap, m_pixmap, m_gcontext, src_x, src_y, 0, 0, w, h);
@@ -257,12 +278,18 @@ void bg_slice::allocate_resources(xcb_visualtype_t* visual) {
   XCB_AUX_ADD_PARAM(&mask, &params, foreground, black_pixel);
   XCB_AUX_ADD_PARAM(&mask, &params, background, black_pixel);
   XCB_AUX_ADD_PARAM(&mask, &params, graphics_exposures, 0);
-  m_connection.pack_values(mask, &params, value_list);
+  connection::pack_values(mask, &params, value_list);
   m_gcontext = m_connection.generate_id();
   m_connection.create_gc(m_gcontext, m_pixmap, mask, value_list.data());
 
   m_log.trace("background_manager: Allocating cairo surface");
   m_surface = make_unique<cairo::xcb_surface>(m_connection, m_pixmap, visual, m_rect.width, m_rect.height);
+
+  /*
+   * Fill pixmap with black in case it is not fully filled by the root pixmap. Otherwise we may render uninitialized
+   * memory
+   */
+  m_connection.poly_fill_rectangle(m_pixmap, m_gcontext, 1, &m_rect);
 }
 
 void bg_slice::free_resources() {

